src/tqec/sketchup/block_graph.py

```diff
@@ -0,0 +1,551 @@
+"""Graph representation of a 3D spacetime defect diagram by explicit blocks."""
+
+from __future__ import annotations
+
+from io import BytesIO
+import typing as ty
+import pathlib
+from dataclasses import dataclass, astuple
+from enum import Enum
+
+import networkx as nx
+
+from tqec.position import Position3D, Direction3D
+from tqec.sketchup.zx_graph import (
+    ZXGraph,
+    NodeType,
+    ZXNode,
+    ZXEdge,
+)
+from tqec.exceptions import TQECException
+
+
+@dataclass(frozen=True)
+class Color3D:
+    """Get face colors along the x, y, and z axes."""
+
+    x: str | None
+    y: str | None
+    z: str | None
+
+    def __post_init__(self) -> None:
+        if any(c not in "xz" for c in astuple(self) if c is not None):
+            raise TQECException("Color must be 'x' or 'z'.")
+
+    def match(self, other: Color3D) -> bool:
+        """Check whether the color matches the other color."""
+        return all(
+            c1 is None or c2 is None or c1 == c2
+            for c1, c2 in zip(astuple(self), astuple(other))
+        )
+
+    def pop_color_at_direction(self, direction: Direction3D) -> Color3D:
+        """Replace the color at the given direction with None."""
+        return self.push_color_at_direction(direction, None)
+
+    def push_color_at_direction(
+        self, direction: Direction3D, color: str | None
+    ) -> Color3D:
+        """Set the color at the given direction."""
+        if direction == Direction3D.X:
+            return Color3D(color, self.y, self.z)
+        elif direction == Direction3D.Y:
+            return Color3D(self.x, color, self.z)
+        else:
+            return Color3D(self.x, self.y, color)
+
+    @staticmethod
+    def from_string(s: str, flip_xz: bool = False) -> "Color3D":
+        s = s.lower()
+        if s == "virtual":
+            return Color3D(None, None, None)
+        if len(s) != 3 or any(c not in "xzo" for c in s):
+            raise TQECException(
+                "s must be a 3-character string containing only 'x', 'z', and 'o'."
+            )
+        colors: list[str | None] = []
+        for c in s:
+            if c == "o":
+                colors.append(None)
+            elif flip_xz:
+                colors.append("z" if c == "x" else "x")
+            else:
+                colors.append(c)
+        return Color3D(*colors)
+
+
+class CubeType(Enum):
+    """Valid cube types in the library."""
+
+    ZXX = "zxx"
+    XZX = "xzx"
+    XXZ = "xxz"
+    XZZ = "xzz"
+    ZXZ = "zxz"
+    ZZX = "zzx"
+    # Virtual cube for open port
+    VIRTUAL = "virtual"
```

Comment:
Equivalent to
```suggestion
class CubeType(StrEnum):
    """Valid cube types in the library."""

    ZXX = auto()
    XZX = auto()
    XXZ = auto()
    XZZ = auto()
    ZXZ = auto()
    ZZX = auto()
    # Virtual cube for open port
    VIRTUAL = auto()
```
per the [documentation of `auto()`](https://docs.python.org/3/library/enum.html#enum.auto). I prefer your version, but the proposed version above removes any possibility of typo in the values.

Reply:
I have thought about using `StrEnum` for that, but it's available only until Python 3.11.

Reply:
Ho, then I am fine with your implementation, sorry I did not realised that it was `>= 3.11`.